gui/src/3D/controls/touchButton3D.ts

```diff
@@ -0,0 +1,387 @@
+
+import { DeepImmutableObject } from "babylonjs/types";
+import { Vector3 } from "babylonjs/Maths/math.vector";
+import { Mesh } from "babylonjs/Meshes/mesh";
+import { AbstractMesh } from "babylonjs/Meshes/abstractMesh";
+import { TransformNode } from "babylonjs/Meshes/transformNode";
+import { Scene } from "babylonjs/scene";
+import { Ray } from "babylonjs/Culling/ray";
+
+import { Button3D } from "./button3D";
+
+/**
+ * Enum for Button States
+ */
+enum ButtonState {
+    /** None */
+    None = 0,
+    /** Pointer Entered */
+    Hover = 1,
+    /** Pointer Down */
+    Press = 2
+}
+
+class TouchButton3DManager {
+    private _touchButtonList = new Map<number, TouchButton3D>();
+    private _buttonIndex = 1;
+
+    private _sceneRegisteredOn: Scene | null;
+    private _collisionHandlingFunction: () => void;
+
+    /**
+     * Creates a new touchButton3DManager
+     */
+    constructor() {
+        let _this = this;
+        this._collisionHandlingFunction = function () {
+            if (_this._sceneRegisteredOn != null) {
+                const touchMeshes = _this._sceneRegisteredOn.getMeshesByTags("touchEnabled");
+
+                _this._touchButtonList.forEach(function (button: TouchButton3D) {
+                    touchMeshes.forEach(function (mesh: Mesh) {
+                        button.collisionCheckForStateChange(mesh);
+                    });
+                });
+            }
+        };
+    }
+
+    public addButton(button: TouchButton3D): number {
+        let index = this._buttonIndex++;
+        this._touchButtonList.set(index, button);
+        return index;
+    }
+
+    public removeButton(index: number): boolean {
+        return this._touchButtonList.delete(index);
+    }
+
+    public enableCollisionHandling(scene: Scene) {
+        if (this._sceneRegisteredOn != scene) {
+            if (this._sceneRegisteredOn != null) {
+                this.disableCollisionHandling();
+            }
+
+            scene.registerBeforeRender(this._collisionHandlingFunction);
+            this._sceneRegisteredOn = scene;
+        }
+    }
+
+    public disableCollisionHandling() {
+        this._sceneRegisteredOn?.unregisterBeforeRender(this._collisionHandlingFunction);
+        this._sceneRegisteredOn = null;
+    }
+}
+
+/**
+ * Class used to create a touchable button in 3D
+ */
+export class TouchButton3D extends Button3D {
+    /** @hidden */
+    //private _buttonState: ButtonState;
+    private _collisionMesh: Mesh;
+    private _collidableFrontDirection: Vector3;
+    private _lastTouchPoint: Vector3;
+
+    private _collidableInitialized = false;
+
+    private _frontOffset = 0;
+    private _backOffset = 0;
+    private _hoverOffset = 0;
+    private _pushThroughBackOffset = 0;
+
+    private _activeInteractions = new Map<number, ButtonState>();
+    private _previousHeight = new Map<number, number>();
+
+    private static _buttonManager = new TouchButton3DManager();

```

Comment:
I just wanted to make it a singleton

Reply:
Why do we want this to be a singleton?

I'm not sure how object construction works in typescript, but it looks like in its current form the TouchButton3DManager is reconstructed every time a button is created. Do we know for sure that all buttons end up using the same singleton with this setup?

---
In reply to: [545277195](https://github.com/BabylonJS/Babylon.js/pull/9620#discussion_r545277195) [](ancestors = 545277195)

Reply:
To be more precise, I wanted something static so that there was a single manager that had knowledge of every button. This is so I can do all of the collision checking within a single function call, allowing me to only call scene.registerBeforeRender once.

I found that having each button register independently caused a masive performance drop in my test scene, since I had 60+ buttons, and found it was directly related to the number of calls to registerBeforeRender.

Though now I'm wondering if I'd still see that drop if I used registerAfterRender...
Maybe someone more knowledgeable on performance costs can tell me what the best way to do this is, given that I'm okay with a frame delay.

Reply:
What happens when there are multiple scenes?

Reply:
I'm not sure. I don't know much about multi-scene setups in babylon.js, but so long as the original scene exists, I imagine this will point to the GUI scene for it. Though now I'm wondering how hand meshes/visuals work across multiple scenes as well, since I believe them to also be attached to the original scene on creation.

Reply:
This is a good question which I need to investigate. I believe you will need to init the hand meshes again for them to work in a new scene. 

Back to the original question, I still don't understand the need of a static private member. removing the static and using _buttonManager should have the same effect, no?

Reply:
@RaananW If I remove the static, then wouldn't each button have their own manager? That would mean multiple calls to scene.registerBeforeRender, which I'm trying to avoid by using a static manager.
I have found that excessive calls to this cause a large performance drop, at least on HoloLens. I noticed this when I copied the button panel playground scene, which generates 60 buttons.